api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -842,3 +842,83 @@ message RequestIDExtension {
   // Request ID extension specific configuration.
   google.protobuf.Any typed_config = 1;
 }
+
+// If this is used, the cluster will only operate on one of the possible upstream protocols (HTTP/1.1, HTTP/2).
+// If :ref:`http2_protocol_options <envoy_api_field_config.cluster.v3.Cluster.http2_protocol_options>` are
+// present, HTTP2 will be used, otherwise HTTP1.1 will be used.
+message ExplicitHttpConfig {
+  oneof protocol_config {
+    config.core.v3.Http1ProtocolOptions http_protocol_options = 1;
+
+    config.core.v3.Http2ProtocolOptions http2_protocol_options = 2;
+  }
+}
+
+// If this is used, the cluster can use either of the configured protocols, and
+// will use whichever protocol was used by the downstream connection.
+message UseDownstreamHttpConfig {
```

Comment:
Also while I split this PR out because ALL THE CONFIGS, it's motivated by the addition of AlpnHttpConfig

  // If this is used for upstream protocol configuration, Envoy will negotiate ALPN to determine if
  // HTTP/1 or HTTP/2 should be used.
  message AlpnHttpConfig {
    config.core.v3.Http1ProtocolOptions http_protocol_options = 3;

    config.core.v3.Http2ProtocolOptions http2_protocol_options = 4;
  }

https://github.com/envoyproxy/envoy/pull/13922/files

I'd like API shephards to be mindful both about how that will work, and how eventual HTTP/3 support will work.
Where for HTTP/2 we can have one cluster, which creates TCP connections which are either HTTP/2 or HTTP/1.1 based on ALPN, HTTP/3 will need to do dual connections on UDP and TCP and so isn't actually configured by ALPN but IMO is going to want to have its own best effort "gimmee the highest protocol I can use" semantics.

That said we already allow creating distinct HTTP/1 and HTTP/2 connection pools with the UseDownstreamHttpConfig configuration, so I think it'll be on par with that.  Just worth keeping in mind to avoid the API deprecation dance in another month or two.

Reply:
How does H3 negotiation work here? There are two connections launched in parallels and UDP preferred if successful? Or does it try UDP and then fall back to TCP (H2)?

Reply:
for new origins, yes, probe in parallel, use what gets established first (semi-random).  Once QUIC is confirmed to work it's assumed to work (maybe modulo network change) until it fails, then dual-probing kicks on again.

I *think* from a config side I'd just configure it as another options in the ALPN config (maybe call it something different) but   I think it's worth the API folks thinking about it as all these refactors land.

Reply:
With regard to ^ I think we should also think about happy eyeballs (dual v6 and v4 connections as well as H3 fallback). I'm guessing we might want to s/AlpnHttpConfig/AutoDetectHttpConfig or something like that? But we can discuss that in the next review?

Reply:
yep, just want folks thinking about that now, so that one lands smoothly as an addition rather than prompting extra reworking